src/server/gc_worker/compaction_filter.rs

```diff
@@ -160,33 +163,75 @@ impl WriteCompactionFilter {
     }
 
     // Do gc before a delete mark.
-    fn gc_before_delete_mark(&mut self, delete: &[u8], prefix: &[u8]) {
-        let mut iter = self.write_iter.take().unwrap();
-        let mut valid = iter.seek(SeekKey::Key(delete)).unwrap()
-            && (iter.key() != delete || iter.next().unwrap());
+    fn handle_deleting(&mut self) {
+        if self.deleting.0.is_empty() {
+            return;
+        }
+
+        let mut valid = self.seek_to_deleting_position();
+        let mut write_iter = self.write_iter.take().unwrap();
         while valid {
-            let (key, value) = (iter.key(), iter.value());
-            let key_prefix = match Key::split_on_ts_for(key) {
-                Ok((key, _)) => key,
-                Err(_) => continue,
-            };
-            if key_prefix != prefix {
+            let (key, value) = (write_iter.key(), write_iter.value());
+            if truncate_ts(key) > self.deleting.0.as_slice() {
+                self.deleting.0.clear();
+                self.deleting.1.clear();
                 break;
             }
-            let write = WriteRef::parse(value).unwrap();
-            if write.short_value.is_none() && write.write_type == WriteType::Put {
-                let def_key = Key::from_encoded_slice(key_prefix).append_ts(write.start_ts);
-                self.delete_default_key(def_key.as_encoded());
+
+            let (mut drain_pos, mut found) = (0, false);
+            while drain_pos < self.deleting.1.len() {
+                match write_iter.key().cmp(self.deleting.1[drain_pos].as_slice()) {
```

Comment:
I thin we can assume both write_iter and `deleting.1` are ordered, and remember `drain_pos` to avoid searching over again for the next key version?

Reply:
`deleting` is drained after every search so there won't be any redundant searches.